### DISCUSSION TITLE
Max doesnt max as expected

### DISCUSSION THREAD

**Discussion by emdete** [fyne-io/fyne#3518] [Q&A] [answered]
i want a layout with a list on top filling up the screen and some buttoms in a horizontal row at the buttom. i use this flow:

```go
	myWindow.SetContent(container.NewVBox(
		container.NewMax(
			widget.NewList(
				func() int {
					return len(data)
				},
				func() fyne.CanvasObject {
					return widget.NewLabel("template")
				},
				func(i widget.ListItemID, o fyne.CanvasObject) {
					o.(*widget.Label).SetText(data[i])
				})),
		container.NewVBox(
			layout.NewSpacer(),
			container.NewHBox(
				layout.NewSpacer(),
				widget.NewButton("Yes", func() {
					log.Println("Yes!")
				}),
				layout.NewSpacer(),
				widget.NewButton("No", func() {
					log.Println("No.")
				}),
				layout.NewSpacer(),
				widget.NewButton("Maybe", func() {
					log.Println("Maybe?")
				}),
				layout.NewSpacer(),
				widget.NewButton("More..", func() {
					log.Println("More..")
				}),
				layout.NewSpacer(),
			))))
```

i expected the Max in the second line would extend the list to the remaining space above the buttons but it doesnt. 

if i replace the VBox in the first line with VSplit it works better, the List ist expanded to half of the screen. If i use Max in the first line the list *is* expanded as well (but fills the whole screen as documented).

what would be the correct way to have the list above (y-wise) the buttons?

**Comment by Jacalz** [comment, marked as answer]
The max layout does not do what you want it to do in this case. A container can only layout its items within its own area, not outside of it. The max layout basically allows you to stack items on top of each other, filling the space they are given, but they will not fill all the remaining space outside of the container.

In this case, I expect that a border layout  is more like what you want. You can place content at the edge positions and then the fifth argument will fill all the remaining space.

I suggest reading through https://developer.fyne.io/explore/layouts and I especially think that you should have a look at https://www.youtube.com/watch?v=LWn1403gY9E as it should help getting a better understanding about how layouts and containers work.

**Reply by emdete** [reply]
sure, i read the document and viewed the vid (while i dont like to watch vids as i cannot search text or cut&paste code from it ;) ) but that did not help to understand how that is meant to work, especially as VSplit or outside Max maxed out the list. The doc warns that always the Min is used but is silent about how to max out remaining space.

the doc of BorderLayout states "NewBorderLayout creates a new BorderLayout instance with top, bottom, left and right objects set. All other items in the container will fill the centre space". it gets 4 parameters (so i wonder you talk about the fifth) and mentions "other items" which i dont know where these are or come from.

can you show a code example on how a border layout could help here?

**Reply by Jacalz** [reply]
> while i dont like to watch vids as i cannot search text or cut&paste code from it

Totally understand. The slides are available here though: https://conf.fyne.io/archive/2021/Layouts.pdf.

> the doc of BorderLayout states "NewBorderLayout creates a new BorderLayout instance with top, bottom, left and right objects set. All other items in the container will fill the centre space". it gets 4 parameters (so i wonder you talk about the fifth) and mentions "other items" which i dont know where these are or come from.

The border layout takes four or more parameters (look at the function signature, especially `objects ...fyne.CanvasObject`), per the documentation: https://pkg.go.dev/fyne.io/fyne/v2/container#NewBorder.
The first four are added at the corresponding edges (at their MinSize) but the fifth (or more after) parameter is placed in the center taking up all the space that the edge items do not fill.

**Reply by Jacalz** [reply]
Ah, I see what you mean now. A problem has crept into the slide about the border layout in that video. Sorry for the confusion.

**Comment by emdete** [comment]
the doc i see says:

```
func NewBorderLayout(top, bottom, left, right fyne.CanvasObject) fyne.Layout
```

no `...` (https://developer.fyne.io/api/v2.3/layout/#func--newborderlayout).

**Reply by Jacalz** [reply]
That must be a bug with the generated API docs. cc @andydotxyz

**Reply by andydotxyz** [reply]
I'm not sure what bug you refer to, can you be more specific?

**Reply by Jacalz** [reply]
You can ignore that. It seems like I was wrong. I thought the link above was for container.NewBorder() but it is for the layout package.

**Reply by Jacalz** [reply]
A container and a layout are different things. I think there might have been some confusion there

**Comment by emdete** [comment]
i think the doc generally needs more links ;) i did not find the other doc you mention..

**Reply by Jacalz** [reply]
The other doc is just the official package documentation site for Go. Here is the general link to the Fyne package there:
https://pkg.go.dev/fyne.io/fyne/v2

**Reply by emdete** [reply]
i started here: https://developer.fyne.io/index.html - how about to point to https://pkg.go.dev/fyne.io/fyne/v2 from there somehow? maybe instead of the "API Documentation"? crosslinking alot helps alot ;)

**Reply by Jacalz** [reply]
It is linked to from the top of https://github.com/fyne-io/fyne/blob/master/README.md and I think the idea is to also have all the API docs available on the developer.fyne.io, hence no link out to pkg.go.dev.

**Reply by andydotxyz** [reply]
Indeed, the api documentation generated in the "api documentation" section of our site is a mirror of the go docs - but with a navigation that aims for shorter pages and a richer menu.

**Comment by emdete** [comment]
ok, i changed my code to:

```
	myWindow.SetContent(container.NewBorder(
		nil, 
		container.NewVBox(
			layout.NewSpacer(),
			container.NewHBox(
				layout.NewSpacer(),
				widget.NewButtonWithIcon("", theme.ConfirmIcon(), func() {
					log.Println("Yes!")
				}),
				layout.NewSpacer(),
				widget.NewButtonWithIcon("", theme.CancelIcon(), func() {
					log.Println("No.")
				}),
				layout.NewSpacer(),
				widget.NewButtonWithIcon("", theme.ContentRedoIcon(), func() {
					log.Println("Maybe?")
				}),
				layout.NewSpacer(),
				widget.NewButtonWithIcon("", theme.MenuIcon(), func() {
					log.Println("More..")
				}),
				layout.NewSpacer(),
			)),
		nil, 
		nil,
		widget.NewList(
			func() int {
				return len(data)
			},
			func() fyne.CanvasObject {
				return widget.NewLabel("template")
			},
			func(i widget.ListItemID, o fyne.CanvasObject) {
				o.(*widget.Label).SetText(data[i])
			}),
		))

```

this does the trick! thank you! :)

**Reply by Jacalz** [reply]
You are welcome. I am not sure though if you need the VBox though. I haven't tested it locally but it might give the same result if you remove it.

**Comment by emdete** [comment]
yes, you are right, that can go too.
